[fix] tooltip text leaking with newlines in help

sfc-gh-lwilby opened the pull request:
## Describe your changes

Fixes tooltip rendering when help text contains newlines (`\n\n`). Previously, the `:help[]` text directive would break, causing help text to render in the markdown/caption content instead of inside the tooltip.

**Solution:**
- Pass help text via react context instead of in the directive.
- Preserves inline tooltip rendering which is necessary for text alignment to work correctly

## GitHub Issue Link (if applicable)

Fixes #13339

## Testing Plan

- [x] E2E Tests

**E2E tests:**
- `e2e_playwright/st_markdown.py` - Test app with tooltips containing newlines
- `e2e_playwright/st_markdown_test.py` - Parameterized tests covering:
  - `st.markdown` with newlines in default alignment
  - `st.caption` with newlines
  - `st.markdown` with center alignment + newlines (verifies design decision)
  - Negative assertions to ensure text doesn't leak into element content

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-lwilby committed:
Fix tooltip text leaking with newlines in help

Escape newlines in help text before embedding in :help[]
directive to prevent text from rendering outside tooltip box.

Fixes #13339

sfc-gh-lwilby committed:
Use sentinel for newline escaping and add unit tests

Replace \\n escape with __STREAMLIT_NEWLINE__ sentinel to avoid
corrupting user text containing literal \n sequences.

Add frontend unit tests for:
- Tooltip with newlines (verifies no text leaking)
- Tooltip with literal backslash-n (verifies no conversion)

sfc-gh-bnisco commented on e2e_playwright/st_markdown.py:
```diff
@@ -494,3 +494,22 @@ def draw_header_test(join_output: bool) -> None:
+    "Center aligned with newlines in tooltip",
+    text_alignment="center",
+    help="Line 1\n\nLine 2\n\nLine 3",
+)
```
**suggestion:** We should add tests for this case, which seems to be broken:

```py
st.markdown("Test", help="Help before ] help after")
```

<img width="234" height="105" alt="Image" src="https://github.com/user-attachments/assets/3b70792c-0001-4d59-80c4-c7d3db21f9a8" />

sfc-gh-bnisco commented on e2e_playwright/st_markdown.py:
```diff
@@ -494,3 +494,22 @@ def draw_header_test(join_output: bool) -> None:
     help="This is a help tooltip!",
 )
+
+# Test for gh-13339: Tooltip with newlines should render correctly
```
**suggestion:** We should add tests for this case, which seems to be broken:

```py
st.markdown("Test", help="Tool tip with \n\n new lines")
```

<img width="93" height="56" alt="Image" src="https://github.com/user-attachments/assets/bda5ae98-dd16-4542-84eb-68887dad8ccf" />

lukasmasuch commented:
@cursor review

sfc-gh-lwilby commented on frontend/lib/src/components/elements/Markdown/Markdown.tsx:
```diff
@@ -75,8 +75,13 @@ function Markdown({ element }: Readonly<MarkdownProps>): ReactElement {
+    // Using a unique sentinel prevents collision with user text containing literal \n.
+    const escapedHelp = help
+      ? help.replace(/\n/g, "__STREAMLIT_NEWLINE__")
+      : ""
```
I think it's less likely the user with use this sentinel than an escaped newline character so that's why we are using this sentinel.

This sentinel doesn't seem like something that would come up in real usage.

sfc-gh-lwilby committed:
Escape closing brackets and fix unicode sentinel

- Escape ] brackets to prevent premature directive closure
- Replace sentinel to avoid markdown special chars (__ for bold)
- Use unicode brackets 〔〕 which have no markdown/HTML meaning
- Add tests for brackets and spaces around newlines

sfc-gh-lwilby committed:
Pass help text via React context instead of directive label

Replace character escaping approach with cleaner context-based solution:
- Use :help[] as marker (empty label)
- Pass help text via React context to CustomHelpIcon
- No escaping needed - all markdown features work naturally

Add comprehensive tests for complex markdown in tooltips:
- Code, links, bold, italic, colors, brackets, emojis
- Snapshot test to verify formatting

sfc-gh-lwilby committed:
Add fallback to children prop in CustomHelpIcon

Keep children prop for developers who manually use :help[content]
directive in their markdown. Context is preferred (from help parameter)
but falls back to children for flexibility.

Document text directive limitations in JSDoc.

sfc-gh-lwilby committed:
Fix test assertions to check actual help content

Remove check for 'st.markdown()' which isn't in the help text.
Add check for 'italic' which is actually present.
Ensures negative assertions catch real text leaking.

sfc-gh-lwilby committed:
Add unit test for empty help text edge case

Verify that help="" doesn't render a tooltip icon
to ensure no unexpected rendering occurs.

sfc-gh-lwilby committed:
update snapshots
